### PR TITLE
GP777 Add message queue to handle associated file msgs

### PR DESCRIPTION
**Pull request by joel-san**
## What does this change?
Adds a message for all the associated files and sends it to rabbitMQ, if its less than 10000 associated files. Solves: [GP-777](https://codemill.atlassian.net/browse/GP-777)

## Accessibility

- [ ] [Tested with screen reader](https://guardian.github.io/source/?path=/docs/docs-06-accessibility--page#screen-readers)
- [ ] [Navigable with keyboard](https://guardian.github.io/source/?path=/docs/docs-06-accessibility--page#keyboard-navigation)
- [ ] [Colour contrast passed](https://guardian.github.io/source/?path=/docs/docs-06-accessibility--page#colour-contrast)
- [ ] [The change doesn't use only colour to convey meaning](https://guardian.github.io/source/?path=/docs/docs-06-accessibility--page#use-of-colour)

### CODE REVIEW

File: common/src/main/scala/com/gu/multimedia/storagetier/framework/MessageProcessingFramework.scala [hunk 1]
```diff
@@ -186,6 +187,49 @@ class MessageProcessingFramework (ingest_queue_name:String,
...
+  }
+  private def internalSendMsg(chan:Channel, outputExchange:String, routingKey:String, stringContent:String) = Try {
+
+    val msgProps = new AMQP.BasicProperties.Builder()
+      .contentType("application/json")
+      .contentEncoding("UTF-8")
+      .messageId(UUID.randomUUID().toString)
+      .build()
+
+    chan.basicPublish(outputExchange, routingKey, false,msgProps, stringContent.getBytes(cs) )
```
Comment (fredex42) [inline common/src/main/scala/com/gu/multimedia/storagetier/framework/MessageProcessingFramework.scala:231]: I am guessing `cs` is the character set, but I can't see where it's defined?

File: common/src/test/scala/com/gu/multimedia/storagetier/framework/MessageProcessingFrameworkSpec.scala [hunk 2]
```diff
@@ -29,6 +29,7 @@ class MessageProcessingFrameworkSpec extends Specification with Mockito {
         ProcessorConfiguration("some-exchange","input.routing.key", "output.routing.key", mockedMessageProcessor)
       )
 
+      val queue = "input-queue"
```
Comment (fredex42) [inline common/src/test/scala/com/gu/multimedia/storagetier/framework/MessageProcessingFrameworkSpec.scala:32]: is this needed?

File: project_restorer/src/main/scala/Main.scala [hunk 3]
```diff
@@ -3,6 +3,7 @@ import Main.retryLimit
 import akka.actor.ActorSystem
 import akka.stream.Materializer
 import com.gu.multimedia.mxscopy.MXSConnectionBuilderImpl
+import com.gu.multimedia.storagetier.framework
```
Comment (fredex42) [inline project_restorer/src/main/scala/Main.scala:6]: does not appear to be neeeded?

File: project_restorer/src/main/scala/PlutoCoreMessageProcessor.scala [hunk 4]
```diff
@@ -43,11 +43,17 @@ class PlutoCoreMessageProcessor(mxsConfig:MatrixStoreConfig)(implicit mat:Materi
    } yield result
   }
 
-  def handleStatusMessage(updateMessage: ProjectUpdateMessage):Future[Either[String, Json]] = {
+
+  def sendToMessageExchange(msgs: Seq[OnlineOutputMessage], routingKey: String, framework: MessageProcessingFramework) = {
+    framework.bulkSendMessages("Nearline.project.e", routingKey, msgs)
+  }
```
Comment (fredex42) [inline project_restorer/src/main/scala/PlutoCoreMessageProcessor.scala:49]: TBH this method is probably not required, you can just call `framework.bulkSendMessages` at line 56 in `handleStatusMessage`

Comment (fredex42) [reply project_restorer/src/main/scala/PlutoCoreMessageProcessor.scala:49]: What does `Nearline.project.e` refer to?

File: common/src/main/scala/com/gu/multimedia/storagetier/framework/MessageProcessingFramework.scala (hunk 1 above)
Comment (joel-san) [reply common/src/main/scala/com/gu/multimedia/storagetier/framework/MessageProcessingFramework.scala:231]: Its declared at line 45

File: project_restorer/src/main/scala/PlutoCoreMessageProcessor.scala (hunk 4 above)
Comment (joel-san) [reply project_restorer/src/main/scala/PlutoCoreMessageProcessor.scala]: I changed it to: `nearline.project-restorer.associatedmsgs` instead. Do you have another name you prefer?

File: common/src/test/scala/com/gu/multimedia/storagetier/framework/MessageProcessingFrameworkSpec.scala (hunk 2 above)
Comment (joel-san) [reply common/src/test/scala/com/gu/multimedia/storagetier/framework/MessageProcessingFrameworkSpec.scala]: Removed it :)

File: common/src/main/scala/com/gu/multimedia/storagetier/framework/MessageProcessingFramework.scala [hunk 5]
```diff
@@ -186,6 +187,49 @@ class MessageProcessingFramework (ingest_queue_name:String,
...
+    })
+  }
+
+
+  def bulkSendMessages[T:io.circe.Encoder](outputExchange:String, routingKey:String, msgs:Seq[T], retry:Int=0) = {
+    withChannel( {chan =>
+      for {
+        _ <- Try { chan.exchangeDeclare(outputExchange, "topic", true)}
+        result <- {val sendResults = msgs.map(_.asJson.noSpaces)
+        .map(stringContent => internalSendMsg(chan, outputExchange, routingKey, stringContent))
```
[the remark this answers is not in the document]
Comment (joel-san) [reply common/src/main/scala/com/gu/multimedia/storagetier/framework/MessageProcessingFramework.scala]: Changed

File: project_restorer/src/main/scala/PlutoCoreMessageProcessor.scala (hunk 4 above)
Comment (fredex42) [reply project_restorer/src/main/scala/PlutoCoreMessageProcessor.scala]: hmm it's the exchange to send the messages on? Personally I would just use the "output exchange" from the framework instance.   The routing key allows us to disambiguate what the messages are from.

**Review by fredex42 — APPROVED** [review]
👍

**Pull request merged**